File: clojure/src/ch1/ex118.clj
;; (load-file "filename.clj")

(defn halve [x] (/ x 2))

(defn mult-invariant [a b acc]
  (cond (= b 0)   acc
        (even? b) (recur (double a) (halve b) acc)
        :else     (recur a (dec b) (+ acc a))))

(defn times [a b]
  (mult-invariant a b 0))


(require
 'clojure.test)

(clojure.test/is (= (times 2 0) 0))
(clojure.test/is (= (times 2 1) 2))
(clojure.test/is (= (times 2 2) 4))
(clojure.test/is (= (times 2 5) 10))
(clojure.test/is (= (times 2 9) 18))
(clojure.test/is (= (times 3 4) 12))
(clojure.test/is (= (times 3 5) 15))

(println 'done)

File: scala/src/ch1/ex118.scala
def even(x: Int) = x % 2 == 0

def mult_inv(b: Int, p: Int, a: Int): Int = p match {
  case 0 => a
  case x if even(x) => mult_inv(b * 2, p / 2, a)
  case _ => mult_inv(b, p - 1, a + b)
}

def mult(b: Int, p: Int) = mult_inv(b, p, 0)

assert(mult(2, 10) == 20)
assert(mult(2, 256) == 512)
assert(mult(3, 4) == 12)
assert(mult(3, 5) == 15)

println("done")
